WIP: ENV-friendly libgit2

tiennou opened the pull request:
This is a work-in-progress implementation of the design I proposed in #4815. For consistency, here's a summary :

The current `git_repository_open_ext` lacks the flexibility we need to protect users from changes to the API w.r.t environment variable we might start supporting. So, this adds `git_repository_open_with_opts`, which supersedes it, and adds a user-configurable whitelist of what we're allowed to customize using the environment. This also modifies how we load values from the environment, so we can apply any of those allowed envvars while the repository is opened (mostly to support `GIT_CONFIG`, as the previous implementation would have already loaded and used the config before we could override it). Opening bare repositories while using the environment is now available.

I'm labeling it as WIP because I didn't write tests for the new behavior yet, as I have a few points worth discussing:

- as per the last commit in the PR, I propose removing the `git_strarray ceiling_dirs` in favor of a `git_strarray`. Since its value depends on the platform-specific path separator, I feel like we're doing a disservice to users by not handling that more transparently. Since `git_repository_open_with_opts` is brand new, now's the time to do it.

- I've discovered that `GIT_REPOSITORY_OPEN_FROM_ENV`'s documentation said this :

> In the future, this flag will also cause `git_repository_open_ext` to
respect $GIT_WORK_TREE and $GIT_COMMON_DIR; currently,
`git_repository_open_ext` with this flag will error out if either
$GIT_WORK_TREE or $GIT_COMMON_DIR is set.

This is now supported, which means I've changed how `git_repository_from_ext` behaves for unsuspecting users, which was the whole point of the implementation (😁). I have no strong opinion on the matter, we just need to choose whether those envvars are used or not w.r.t to backward-compat.

- I tried to minimize changes to the "find_repo" function, but I haven't checked for work that might be done for nothing, in case the environment specifies everything. IMHO it would not be worthwhile.

- Documentation-checking. From the top of my head, the semantics of what should happen when `ceiling_dirs` is specified as both a parameter and its envvar form. I'm leaning toward using the more explicit value (ie. listening to the code vs. the environment) in that case.

pks-t commented on src/repository.c:
```diff
@@ -429,6 +429,73 @@ static int read_gitfile(git_buf *path_out, const char *file_path)
+			goto cleanup;
+
+		end = git_buf_cstr(&alts_buf) + git_buf_len(&alts_buf);
+		for (sep = alt = alts_buf.ptr; sep != end; alt = sep+1) {
```
Spaces missing around `+`

pks-t commented on src/repository.c:
```diff
@@ -598,94 +665,90 @@ int git_repository_open_bare(
+	struct repo_env *env,
 	const char *start_path,
-	const git_repository_open_options *opts)
+	git_repository_open_options *opts)
```
So instead of opening the repository, we now just try to find all necessary paths and return that instead? Shouldn't the function be renamed in that case?

tiennou replied:
Fixed, I've renamed two offending functions to clarify what's being setup from what (the new `struct repo_env`, and this one).

pks-t commented on src/repository.c:
```diff
@@ -598,94 +665,90 @@ int git_repository_open_bare(
-			flags |= GIT_REPOSITORY_OPEN_NO_SEARCH;
-			flags |= GIT_REPOSITORY_OPEN_NO_DOTGIT;
+			opts->flags |= GIT_REPOSITORY_OPEN_NO_SEARCH;
+			opts->flags |= GIT_REPOSITORY_OPEN_NO_DOTGIT;
```
Honestly, it's a bit weird to modify the flags passed in by callers

pks-t commented on src/repository.c:
```diff
@@ -867,11 +857,19 @@ int git_repository_open_with_opts(
 
+	if (use_env && git_buf_len(&env.workdir_path)) {
+		git_buf_swap(&workdir, &env.workdir_path);
+	}
```
Useless braces

pks-t commented on include/git2/repository.h:
```diff
@@ -193,6 +193,9 @@ typedef struct {
 	 * the search for a containing repository should terminate.
+	 *
+	 * FIXME: this depends on the value of GIT_PATH_LIST_SEPARATOR, which is
+	 * platform-specific. Make that a git_strarray ?
```
I think it's perfectly fine to fix this API while we're at it. The old function could simply split the string it got into the string array

tiennou commented on src/repository.c:
```diff
@@ -618,7 +619,7 @@ static int _git_repository_open_ext_from_env(
 
 	if (!start_path) {
-		error = git__getenv(&dir_buf, "GIT_DIR");
+		error = git__getenv_wl(&dir_buf, "GIT_DIR", &opts->allowed_env);
```
`allowed_env` is initialized by either the backward-compatibility code in `git_repository_open_ext`, or explicitly by the user via the new options struct, which defaults to `NULL`. So unless I'm missing something this *should* behave as expected.

tiennou replied:
Disregard that, I mistook pointers for non-pointers 🙄.

tiennou commented on include/git2/repository.h:
```diff
@@ -193,6 +199,14 @@ typedef struct {
+	 * When GIT_REPOSITORY_OPEN_FROM_ENV is set, this can be used to
+	 * whitelist the environment variables we're allowed to access.
+	 *
+	 * An empty list means that we can use any of them.
```
Sounds like I've confused myself too : a `NULL` whitelist means "any", an empty one means "none". I fail to see the use case for "empty" so I'm not sure if it makes sense. I'll fix & rephrase though.

tiennou committed:
repo: introduce git_repository_open_with_opts

tiennou committed:
repo: ensure the current _open_ext semantics are preserved

tiennou committed:
util: add a whitelistable getenv helper

tiennou committed:
repo: add a list of envvars we're allowed to use when opening

The goal is to make the list of allowed envvars user-specified, so users
can specify what we're allowed to use from the environment.

This also freezes the current list uses when opening in `_open_ext`, for
backward-compatibility reasons.

tiennou committed:
repo: obey GIT_WORK_TREE/GIT_COMMON_DIR envvars

tiennou committed:
repo: support GIT_CONFIG config file override

tiennou committed:
buf: add git_buf_slice

tiennou committed:
repo: ODB alternates are also a PATHLIST_SEPARATED list

tiennou committed:
strarray: add a splitter for PATH_LIST-separated strings

tiennou committed:
repo: make ceiling_dirs a strarray

tiennou committed:
repo: modify how the environment is loaded and applied

Previously, when `GIT_REPOSITORY_OPEN_FROM_ENV` was used, we'd end up
calling `git_repository_open_ext` recursively with another set of flags
and ceiling_dirs.

This prevents us from supporting some environment variables (eg. GIT_CONFIG)
because part of `git_repository_open_ext` would touch the config before
we'd get a chance to tweak it. AFAICT `GIT_WORK_TREE` and `GIT_COMMON_DIR`
might have encountered the same problem.

Values from the environment are now stored in a struct, so those can
applied as soon as possible while the repository is being setup.

tiennou committed:
WIP: investigate

It seems the find-workdir-from-gitdir was broken.
The first find_env will setup something, which the rest of the code
will trample over.

tiennou committed:
cmake: correct the link stanza for CoreFoundation

LIBRARIES is the (absolute?) path to the library.
LDFLAGS is the full linker stanza to correctly link with this lib.

By passing LIBRARIES as LIBGIT_LIBS, the linker ends up with the
absolute path for the SDK'ed version of CoreFoundation (which doesn't
exist), instead of the familiar `-framework CoreFoundation`.

tiennou committed:
WIP: debugging

tiennou commented on src/repository.c:
```diff
@@ -485,19 +535,22 @@ static int find_repo(
+					git_buf_set(&env->gitdir_path, path.ptr, path.size);
+					GIT_ERROR_CHECK_ALLOC_BUF(&env->gitdir_path);
+
+					git_buf_attach(&env->gitlink_path,
```
This reads a (potentially non-existing) `$GIT_DIR/gitdir` file in `gitlink`.

tiennou commented on src/repository.c:
```diff
@@ -529,31 +582,37 @@ static int find_repo(
+			git_path_dirname_r(&env->workdir_path, path.ptr);
+			git_path_to_dir(&env->workdir_path);
+		}
+		if (git_buf_oom(&env->workdir_path)) {
```
Here, we set `workdir_path` as whatever `dirname(path)` is.

tiennou commented on src/repository.c:
```diff
@@ -790,48 +843,87 @@ static int repo_is_worktree(unsigned *out, const git_repository *repo)
 		GIT_ERROR_CHECK_ALLOC(repo->commondir);
 	}
 
+//	if (git_buf_len(&env.workdir_path) > 0) {
```
I've added this, since it seemed to be provided/detected. This causes a crash further down, hence the commenting.

tiennou commented on src/repository.c:
```diff
@@ -790,48 +843,87 @@ static int repo_is_worktree(unsigned *out, const git_repository *repo)
+		GIT_ERROR_CHECK_ALLOC(repo->configfile);
+	}
+
 	if ((error = repo_is_worktree(&is_worktree, repo)) < 0)
```
[Function](https://github.com/libgit2/libgit2/blob/master/src/repository.c#L770).

This could have easily been part of `repo_env_find`, since it only look at the filesystem.

tiennou commented on src/repository.c:
```diff
@@ -846,32 +938,66 @@ int git_repository_open_ext(
 		if (config &&
 		    ((error = load_config_data(repo, config)) < 0 ||
-		     (error = load_workdir(repo, config, &workdir)) < 0))
+		     (error = load_workdir(repo, config, &env.workdir_path)) < 0))
```
We move on to actually setup the working directory, at https://github.com/libgit2/libgit2/pull/4856/files#diff-7cbc97c802e21b199f176b493cec8e9cR300, and we pass the workdir identified by the "find" step. This is the part where the crash I'm getting comes from, and I'm starting to get completely lost (and I can't inline-review that code anyway, so there goes) :

- `core.worktree` is loaded from the config, as it can now be made use of.
- if `repo->is_worktree` is true (it only performed filesystem-checks), we read `repo->gitdir` + "gitdir" (already loaded while finding in `env->gitlink`), `dirname()` that, and use the result as our workdir.
- if `core.worktree` is set, we check its value and use that as our workdir.
- if the `parent_path` is passed, we check that it's a directory and use that.
- finally, we `dirname(repo->gitdir)` and use that.

It seems that `core.worktree` gets a say iif the repository structure is slightly "wrong" — a `$GIT_DIR/gitdir` file that doesn't point to a valid repo. The fact that it is passing as a parameter the very thing this code should be setting up doesn't help 😉.

tiennou reviewed (commented):
Rebased, as I've wanted to try to polish it a little, and hit a snag. This is still WIP, but the current code really needs a refactor, and I don't think I can handle it without re-reading a whole lot of git manpages, so…

This should be green, but as I had to comment "innocuous looking" code to fix it, I had to look more closely at what was going on. I've left review comments about those issues, in case someone else has any insight. Note that there are a bunch of missing oom checks in there, which I didn't attempt to fix yet.

ethomson commented:
Tidying up some old PRs. We support these environment variables now.
